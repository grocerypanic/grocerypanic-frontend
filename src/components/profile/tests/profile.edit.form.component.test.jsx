import { render, within, fireEvent, waitFor } from "@testing-library/react";
import React from "react";
import Strings from "../../../configuration/strings";
import ProfileForm, { nullFunction } from "../profile.edit.form.component";

jest.mock("../../../configuration/theme", () => {
  return {
    ...jest.requireActual("../../../configuration/theme"),
    ui: {
      alertTimeout: 10,
    },
  };
});

const mockUser = {
  first_name: "Niall",
  last_name: "Byrne",
  timezone: "mockTime1",
  id: 0,
  has_profile_initialized: false,
};

const props = {
  user: [mockUser],
  title: "MockTitle",
  helpText: "MockHelpText",
  transaction: false,
  timezones: [
    { id: 0, name: "mockTime1" },
    { id: 1, name: "mockTime2" },
  ],
  handleSave: jest.fn(),
};

let utils;

describe("Setup Environment for ProfileForm test", () => {
  let currentProps;
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const renderHelper = (currentProps, command = render) => {
    utils = command(<ProfileForm {...currentProps} />);
  };

  describe("outside of a transaction", () => {
    describe("with a valid timezone", () => {
      beforeEach(async () => {
        currentProps = { ...props };
        currentProps.transaction = false;
        renderHelper(currentProps);
      });

      it("should match the snapshot on file (styles)", () => {
        expect(utils.container).toMatchSnapshot();
      });

      it("populates the hint dialogue correctly", () => {
        utils.getByText(currentProps.helpText);
      });

      it("populates the first name correctly", () => {
        const element = utils.getByTestId("input_first_name");
        expect(element.value).toBe(mockUser.first_name);
      });

      it("populates the last name correctly", () => {
        const element = utils.getByTestId("input_last_name");
        expect(element.value).toBe(mockUser.last_name);
      });

      it("populates timezone correctly", () => {
        const element = utils.getByTestId("input_timezone_parent");
        const dropdown = element.firstChild.querySelector(
          ".list__dropdown-indicator"
        );
        fireEvent.mouseDown(dropdown, "click");
        const menu = within(element.firstChild.querySelector(".list__menu"));

        menu.getByText(currentProps.timezones[0].name);
        menu.getByText(currentProps.timezones[1].name);
      });

      it("creates the button correctly", () => {
        const element = utils.getByTestId("submit");
        expect(element.innerHTML).toBe(Strings.Profile.SaveButton);
        expect(element).toHaveClass("btn-success");
      });

      it("handles save as expected correctly, valid data", async () => {
        const alert = within(utils.getByTestId("alert"));
        expect(alert.queryByText(Strings.Profile.SaveAction)).toBeNull();

        const element = utils.getByTestId("submit");
        fireEvent.click(element, "click");

        expect(currentProps.handleSave).toBeCalledTimes(1);
        expect(currentProps.handleSave).toBeCalledWith({
          ...mockUser,
          has_profile_initialized: true,
        });

        expect(alert.queryByText(Strings.Profile.SaveAction)).toBeTruthy();
        await waitFor(() =>
          expect(alert.queryByText(Strings.Profile.SaveAction)).toBeNull()
        );
      });
    });
  });
  describe("during a transaction", () => {
    beforeEach(async () => {
      currentProps = { ...props };
      currentProps.transaction = true;
      renderHelper(currentProps);
    });

    it("should match the snapshot on file (styles)", () => {
      expect(utils.container).toMatchSnapshot();
    });

    it("creates the button correctly", () => {
      const element = utils.getByTestId("submit");
      expect(element.innerHTML).toBe(Strings.Profile.SaveButton);
      expect(element).toHaveClass("btn-secondary");
    });

    it("handles save as expected correctly", () => {
      const alert = within(utils.getByTestId("alert"));
      expect(alert.queryByText(Strings.Profile.SaveAction)).toBeNull();

      const element = utils.getByTestId("submit");
      fireEvent.click(element, "click");

      expect(currentProps.handleSave).toBeCalledTimes(0);
    });
  });
});

describe("Setup Environment for nullFunction test", () => {
  it("should be a noop", () => {
    nullFunction();
  });
});
