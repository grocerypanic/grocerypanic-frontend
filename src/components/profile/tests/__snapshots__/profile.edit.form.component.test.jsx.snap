// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Setup Environment for ProfileForm test during a transaction should match the snapshot on file (styles) 1`] = `
.c3 {
  list-style-type: none;
  list-style-position: outside;
  padding-left: 0px;
  margin: 0px;
  min-width: 200px;
  background: #808080;
  padding: 10px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #FFFFFF;
  margin: 5px;
  padding: 5px;
  min-width: 220px;
  max-width: 350px;
}

.c4 .form-group {
  margin-bottom: 2px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 .btn {
  margin-top: 20px;
}

.c2 {
  width: 100%;
  margin-left: 10px;
  margin-right: 10px;
  text-align: center;
}

.c0 {
  margin-top: calc(50px + 26px);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .nav-tabs {
  height: 28px !important;
  font-size: 0.75em;
}

.c0 a[role="tab"] {
  padding-bottom: 0px;
  padding-top: 4px;
  padding-left: 8px;
  padding-right: 8px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c6 .alert-light {
  padding-top: 5px;
  padding-bottom: 0px;
  margin-bottom: 0px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-top: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 0.7em;
  color: #000000;
}

.c8 {
  text-align: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2 alert alert-success"
      >
        MockTitle
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <form
            class=""
          >
            <div
              class="row form-group"
            >
              <div
                class="col-2 col-form-label"
              />
              <div
                class="col-12"
              >
                <small
                  class="text-muted form-text"
                >
                  What's your first name?
                </small>
              </div>
              <div
                class="col-12"
              >
                <input
                  class="form-control form-control-20"
                  data-testid="input_first_name"
                  id="first_name"
                  maxlength="150"
                  minlength="2"
                  name="first_name"
                  placeholder=""
                  readonly=""
                  type="text"
                  value="Niall"
                />
              </div>
            </div>
            <div
              class="row form-group"
            >
              <div
                class="col-2 col-form-label"
              />
              <div
                class="col-12"
              >
                <small
                  class="text-muted form-text"
                >
                  And your last name?
                </small>
              </div>
              <div
                class="col-12"
              >
                <input
                  class="form-control form-control-20"
                  data-testid="input_last_name"
                  id="last_name"
                  maxlength="150"
                  minlength="2"
                  name="last_name"
                  placeholder=""
                  readonly=""
                  type="text"
                  value="Byrne"
                />
              </div>
            </div>
            <div
              class="row form-group"
            >
              <div
                class="col-2 col-form-label"
              />
              <div
                class="col-12"
              >
                <small
                  class="text-muted form-text"
                >
                  What timezone do you live in?
                </small>
              </div>
              <div
                class="col-12"
                data-testid="input_timezone_parent"
              >
                <div
                  class=" css-2b097c-container"
                >
                  <span
                    aria-atomic="false"
                    aria-live="polite"
                    aria-relevant="additions text"
                    class="css-1f43avz-a11yText-A11yText"
                  />
                  <div
                    class="list__control css-yk16xz-control"
                  >
                    <div
                      class="list__value-container list__value-container--has-value css-g1d714-ValueContainer"
                    >
                      <div
                        class="list__single-value css-1uccc91-singleValue"
                      >
                        mockTime1
                      </div>
                      <div
                        class="css-b8ldur-Input"
                      >
                        <div
                          class="list__input"
                          style="display: inline-block;"
                        >
                          <input
                            aria-autocomplete="list"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            id="react-select-9-input"
                            spellcheck="false"
                            style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="list__indicators css-1hb7zxy-IndicatorsContainer"
                    >
                      <div
                        aria-hidden="true"
                        class="list__indicator list__loading-indicator css-at12u2-loadingIndicator"
                      >
                        <span
                          class="css-1yvy2vo-LoadingDot"
                        />
                        <span
                          class="css-zoievk-LoadingDot"
                        />
                        <span
                          class="css-x748d8-LoadingDot"
                        />
                      </div>
                      <span
                        class="list__indicator-separator css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class="list__indicator list__dropdown-indicator css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="c5"
            >
              <button
                class="btn btn-secondary"
                data-testid="submit"
                type="submit"
              >
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      class="c6"
    >
      <div
        class="alert alert-light"
        data-testid="alert"
        role="alert"
      >
         
      </div>
    </div>
    <div
      class="c7"
      data-testid="HintDialogue"
    >
      <div
        class="c8"
      >
        MockHelpText
      </div>
    </div>
  </div>
</div>
`;

exports[`Setup Environment for ProfileForm test outside of a transaction with a valid timezone should match the snapshot on file (styles) 1`] = `
.c3 {
  list-style-type: none;
  list-style-position: outside;
  padding-left: 0px;
  margin: 0px;
  min-width: 200px;
  background: #808080;
  padding: 10px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #FFFFFF;
  margin: 5px;
  padding: 5px;
  min-width: 220px;
  max-width: 350px;
}

.c4 .form-group {
  margin-bottom: 2px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 .btn {
  margin-top: 20px;
}

.c2 {
  width: 100%;
  margin-left: 10px;
  margin-right: 10px;
  text-align: center;
}

.c0 {
  margin-top: calc(50px + 26px);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .nav-tabs {
  height: 28px !important;
  font-size: 0.75em;
}

.c0 a[role="tab"] {
  padding-bottom: 0px;
  padding-top: 4px;
  padding-left: 8px;
  padding-right: 8px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c6 .alert-light {
  padding-top: 5px;
  padding-bottom: 0px;
  margin-bottom: 0px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-top: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 0.7em;
  color: #000000;
}

.c8 {
  text-align: center;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2 alert alert-success"
      >
        MockTitle
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <form
            class=""
          >
            <div
              class="row form-group"
            >
              <div
                class="col-2 col-form-label"
              />
              <div
                class="col-12"
              >
                <small
                  class="text-muted form-text"
                >
                  What's your first name?
                </small>
              </div>
              <div
                class="col-12"
              >
                <input
                  class="form-control form-control-20"
                  data-testid="input_first_name"
                  id="first_name"
                  maxlength="150"
                  minlength="2"
                  name="first_name"
                  placeholder=""
                  type="text"
                  value="Niall"
                />
              </div>
            </div>
            <div
              class="row form-group"
            >
              <div
                class="col-2 col-form-label"
              />
              <div
                class="col-12"
              >
                <small
                  class="text-muted form-text"
                >
                  And your last name?
                </small>
              </div>
              <div
                class="col-12"
              >
                <input
                  class="form-control form-control-20"
                  data-testid="input_last_name"
                  id="last_name"
                  maxlength="150"
                  minlength="2"
                  name="last_name"
                  placeholder=""
                  type="text"
                  value="Byrne"
                />
              </div>
            </div>
            <div
              class="row form-group"
            >
              <div
                class="col-2 col-form-label"
              />
              <div
                class="col-12"
              >
                <small
                  class="text-muted form-text"
                >
                  What timezone do you live in?
                </small>
              </div>
              <div
                class="col-12"
                data-testid="input_timezone_parent"
              >
                <div
                  class=" css-2b097c-container"
                >
                  <span
                    aria-atomic="false"
                    aria-live="polite"
                    aria-relevant="additions text"
                    class="css-1f43avz-a11yText-A11yText"
                  />
                  <div
                    class="list__control css-yk16xz-control"
                  >
                    <div
                      class="list__value-container list__value-container--has-value css-g1d714-ValueContainer"
                    >
                      <div
                        class="list__single-value css-1uccc91-singleValue"
                      >
                        mockTime1
                      </div>
                      <div
                        class="css-b8ldur-Input"
                      >
                        <div
                          class="list__input"
                          style="display: inline-block;"
                        >
                          <input
                            aria-autocomplete="list"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            id="react-select-2-input"
                            spellcheck="false"
                            style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                    </div>
                    <div
                      class="list__indicators css-1hb7zxy-IndicatorsContainer"
                    >
                      <span
                        class="list__indicator-separator css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class="list__indicator list__dropdown-indicator css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="c5"
            >
              <button
                class="btn btn-success"
                data-testid="submit"
                type="submit"
              >
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      class="c6"
    >
      <div
        class="alert alert-light"
        data-testid="alert"
        role="alert"
      >
         
      </div>
    </div>
    <div
      class="c7"
      data-testid="HintDialogue"
    >
      <div
        class="c8"
      >
        MockHelpText
      </div>
    </div>
  </div>
</div>
`;
